File: tools/trim_audio.cpp
#include <filesystem>
#include <cstdlib>
#include <iostream>

void trim_audio(const std::string& input_path, const std::string& output_path, int start_sec, int duration_sec) {
    // Check if input file exists
    if (!std::filesystem::exists(input_path)) {
        std::cerr << "Input file not found: " << input_path << std::endl;
        return;
    }

    // Build FFmpeg command
    std::string command = "ffmpeg -y -i " + input_path +
                          " -ss " + std::to_string(start_sec) +
                          " -t " + std::to_string(duration_sec) +
                          " " + output_path;

    // Execute the command
    int ret = std::system(command.c_str());
    if (ret == 0) {
        std::cout << "Trimmed audio saved to " << output_path << std::endl;
    } else {
        std::cerr << "Error trimming audio!" << std::endl;
    }
}

int main() {
    trim_audio("./examples/audio/dev_audio/itmightcomeback.wav", "./examples/audio/test_audio/trimmed_audio.wav", 0, 5);
    return 0;
}

File: tools/trim_audio.py
from pydub import AudioSegment
from pydub.utils import mediainfo
import os

def trim_audio(input_path, output_path, start_ms=0, end_ms=5000, output_format="wav"):
    """
    Trim an audio file to the specified range and save it efficiently.
    
    Args:
        input_path (str): Path to the input audio file.
        output_path (str): Path to save the trimmed audio.
        start_ms (int): Start time in milliseconds.
        end_ms (int): End time in milliseconds.
        output_format (str): Format for the output file (default is "wav").
    """
    try:
        # Check if the input file exists
        if not os.path.exists(input_path):
            raise FileNotFoundError(f"Input file {input_path} not found.")

        # Ensure the end time is not greater than the file duration
        info = mediainfo(input_path)
        duration_ms = float(info["duration"]) * 1000  # Convert seconds to milliseconds
        if end_ms > duration_ms:
            raise ValueError(f"End time exceeds audio duration ({duration_ms} ms).")

        # Load only the required segment
        duration = (end_ms - start_ms) / 1000  # Convert to seconds
        audio = AudioSegment.from_file(input_path, start_second=start_ms / 1000, duration=duration)
        
        # Export the trimmed audio
        audio.export(output_path, format=output_format)
        print(f"Trimmed audio saved to {output_path}")

    except Exception as e:
        print(f"Error trimming audio: {e}")

# Example usage
trim_audio(
    "./examples/audio/dev_audio/itmightcomeback.wav",
    "./examples/audio/dev_audio/trimmed_audio.wav",
    start_ms=0,
    end_ms=5000
)
